lib/index.js

```diff
@@ -28,26 +43,53 @@ NamedURLResolverClass.prototype.escape = function(string) {
 
 NamedURLResolverClass.prototype.resolve = function(name, params) {
     if(name && (name in this.routesMap)) {
-        if(!params) params = {};
-
         var routePath = this.routesMap[name];
-        for(var paramName in params) {
-            if(params.hasOwnProperty(paramName)) {
-                let paramRegex = new RegExp('(/|^):' + paramName + '(/|$)');
-                let paramValue = this.escape('' + params[paramName]);
-                routePath = routePath.replace(paramRegex, '$1' + paramValue + '$2');
+
+        if(!params) {
+            return routePath;
+        }
+
+        for (var paramName in params) {
+            if (params.hasOwnProperty(paramName)) {
+                var paramValue = params[paramName];
+        
+                if (paramName === "splat") { // special param name in RR, used for "*" and "**" placeholders
+                    paramValue = toArray(paramValue); // when there are multiple globs, RR defines "splat" param as array.
+                    var i = 0;
+                    routePath = routePath.replace(reSplatParams, (match) => {
+                        var val = paramValue[i++];
+                        // don't escape slashes for double star, as "**" considered greedy by RR spec
+                        return val == null ? "" : this.escape('' + val, match === "*");
+                    });
+                } else {
+                    // Rougly resolve all named placeholders.
+                    // Cases: 
+                    // - "/path/:param"
+                    // - "/path/(:param)"
+                    // - "/path(/:param)"
+                    // - "/path(/:param/):another_param"
+                    // - "/path/:param(/:another_param)"
+                    // - "/path(/:param/:another_param)"
+                    var paramRegex = new RegExp('(\/|\\(|\\)|^):' + paramName + '(\/|\\)|\\(|$)');
+                    var replacement = '$1' + this.escape('' + paramValue) + '$2';
+                    routePath = routePath.replace(paramRegex, replacement);
+                }
             }
         }
-        return routePath;
+        return routePath
+            // Remove braces around resolved optional params (i.e. "/path/(value)")
```

Comment:
Totally agreed that a full solution would be an overkill. 

Actually, talking about nesting, I meant only that problem, that inserted values with parentheses will be then stripped out by `reResolvedOptionalParams`, nothing more. Didn't even thought about nested optional params, because I can't imagine a usecase when this would be really required. 

For example, in your case `/user(/:teamId(/:projectId(/:userId)))`, which is good real-world example, one route will display a lot of different data, and it's unlikely that there will be just one handler component for this. Probably, in real world it will look like this:
```jsx
<Route path="/user" component={UserPage}>
   <Route path=":teamId" component={TeamPage}> 
       <Route path=":projectId" component={TeamProjectPage}> 
           <Route path=":userId" component={...etc...} /> 
       </Route>
   </Route>
</Route>
```
That's the tree you spoke about. Explicit and without any regexp black magic.
So I hope that in practice there never will be a need in nested optional params. I don't even know exactly, whether RR really allows and understands them.